src/plugins/profiling/common/callercallee.ts

```diff
@@ -98,107 +97,145 @@ function sortRelevantTraces(relevantTraces: Map<StackTraceID, relevantTrace>): S
-  SourceLine: 0,
-  Samples: 0,
 };
 
-export function buildCallerCalleeNode(node: Partial<CallerCalleeNode> = {}): CallerCalleeNode {
-  return override(defaultCallerCalleeNode, node);
+export function buildCallerCalleeNode(partial: Partial<CallerCalleeNode> = {}): CallerCalleeNode {
+  const node = {} as CallerCalleeNode;
+
+  node.Callers = clone(partial.Callers ?? []);
+  node.Callees = clone(partial.Callees ?? []);
+  node.FileID = partial.FileID ?? '';
+  node.FrameType = partial.FrameType ?? 0;
+  node.ExeFileName = partial.ExeFileName ?? '';
+  node.FunctionID = partial.FunctionID ?? '';
+  node.FunctionName = partial.FunctionName ?? '';
+  node.AddressOrLine = partial.AddressOrLine ?? 0;
+  node.FunctionSourceLine = partial.FunctionSourceLine ?? 0;
+  node.FunctionSourceID = partial.FunctionSourceID ?? '';
+  node.FunctionSourceURL = partial.FunctionSourceURL ?? '';
+  node.SourceFilename = partial.SourceFilename ?? '';
+  node.SourceLine = partial.SourceLine ?? 0;
+  node.Samples = partial.Samples ?? 0;
+
+  return node;
 }
 
 // selectCallerCalleeData is the "standard" way of merging multiple frames into
 // one node. It simply takes the data from the first frame.
 function selectCallerCalleeData(frameMetadata: Set<StackFrameMetadata>, node: CallerCalleeNode) {
   for (const metadata of frameMetadata) {
+    node.FileID = metadata.FileID;
+    node.FrameType = metadata.FrameType;
     node.ExeFileName = metadata.ExeFileName;
     node.FunctionID = metadata.FunctionName;
     node.FunctionName = metadata.FunctionName;
+    node.AddressOrLine = metadata.AddressOrLine;
+
+    // Unknown/invalid offsets are currently set to 0.
+    //
+    // In this case we leave FunctionSourceLine=0 as a flag for the UI that the
+    // FunctionSourceLine should not be displayed.
+    //
+    // As FunctionOffset=0 could also be a legit value, this work-around needs
+    // a real fix. The idea for after GA is to change FunctionOffset=-1 to
+    // indicate unknown/invalid.
+    if (metadata.FunctionOffset > 0) {
+      node.FunctionSourceLine = metadata.SourceLine - metadata.FunctionOffset;
+    } else {
+      node.FunctionSourceLine = 0;
+    }
+
     node.FunctionSourceID = metadata.SourceID;
     node.FunctionSourceURL = metadata.SourceCodeURL;
-    node.FunctionSourceLine = metadata.FunctionLine;
-    node.SourceLine = metadata.SourceLine;
-    node.FrameType = metadata.FrameType;
     node.SourceFilename = metadata.SourceFilename;
-    node.FileID = metadata.FileID;
-    node.AddressOrLine = metadata.AddressOrLine;
+    node.SourceLine = metadata.SourceLine;
     break;
   }
 }
 
 function sortNodes(
   nodes: Map<FrameGroupID, CallerCalleeIntermediateNode>
 ): CallerCalleeIntermediateNode[] {
-  const sortedNodes: CallerCalleeIntermediateNode[] = new Array(nodes.size);
+  const sortedNodes = new Array<CallerCalleeIntermediateNode>();
   for (const node of nodes.values()) {
-    sortedNodes.push(node);
+    sortedNodes.push(clone(node));
```

Comment:
why do we need the `clone` here ? Isn't all the data read-only after initial creation ?

Reply:
You're right, this is unnecessary so I removed it.

For background, I had to clone some of the objects to prevent circular references in the intermediate representation. This was an issue when the final callgraph was converted to JSON (the Kibana server would crash after running out of memory).